File: Programming/Lv.2/가장 큰 수/solution.py
# https://school.programmers.co.kr/learn/courses/30/lessons/42746#
def quick_sort(arr):
    if len(arr) <= 1:
        return arr
    
    pivot = str(arr[len(arr) // 2])
    lesser_arr, equal_arr, greater_arr = [], [], []
    
    for s in arr:
        s = str(s)
        if s + pivot < pivot + s:
            greater_arr.append(s)
        elif s + pivot > pivot + s:
            lesser_arr.append(s)
        else:
            equal_arr.append(s)
    
    return quick_sort(lesser_arr) + equal_arr + quick_sort(greater_arr)


def solution(numbers):
    return str(int(''.join(str(s) for s in quick_sort(numbers))))

File: Programming/Lv.2/가장 큰 수/solution.js
// https://school.programmers.co.kr/learn/courses/30/lessons/42746?language=javascript#
function quickSort(arr) {
  if (arr.length <= 1) {
    return arr;
  }

  const pivot = arr[Math.floor(arr.length / 2)].toString();
  const lessers = [];
  const equals = [];
  const greaters = [];

  arr.forEach((s) => {
    const v = s.toString();

    if (v + pivot < pivot + v) {
      greaters.push(v);
    } else if (v + pivot > pivot + v) {
      lessers.push(v);
    } else {
      equals.push(v);
    }
  });

  return [...quickSort(lessers), ...equals, ...quickSort(greaters)];
}

function solution(numbers) {
  // JS에서 int로 표현하는 데에 한계가 있으므로, 이를 유념해야 한다. 무조건 int로 변환하면 오차가 발생한다.
  const answer = quickSort(numbers).join("");
  if (answer.split("").filter((s) => s !== "0").length === 0) {
    return "0";
  } else {
    return answer;
  }
}
